Ncvetkovic/official docs to repo (#298)
tenstorrent/tt-llk: docs/llk/l1/intro.md

# Introduction to Tenstorrent's Low-Level Kernels (LLKs)

## Overview

> Every Tenstorrent chip consists of a grid of Tensix Cores that work together to run advanced ML models.

Reviewer: ```suggestion
Each Tenstorrent chip consists of a grid of Tensix Cores that work together to run advanced ML models.
```

Tenstorrent's Low-Level Kernels (LLKs) are essential software components of our AI ecosystem.

> These kernels enable AI model operation on Tensix cores at high performance, acting as the first software
> layer above the hardware, and providing direct access to its computational capabilities.

Reviewer: ```suggestion
These kernels enable AI model operation on Tensix cores at high performance. They act as the first software layer above the hardware and provide direct access to its computational capabilities.
```
The sentence is a bit long. Consider splitting for clarity.

Developers use low-level kernels and primitives to efficiently build complex operations.

![Tensix Core to Chip](images/LLK-L1-Tensix-1-25_official.png)

## Purpose and Functionality

LLKs control the Tensix Engine, the fundamental computational unit of the Tensix Core, using the custom Tensix instruction set (Tensix ISA) extensions. LLKs also provide an easy way to use the Tensix ISA, allowing operations to be performed with maximum efficiency and fully utilizing the hardware's parallelism and optimized compute resources.

As the foundational software layer, LLKs provide a direct interface for higher-level applications to harness the full power of the Tensix core, encapsulating hardware specifics into software primitives. They enable low-level tasks like data movement and mathematical operations to be executed at peak performance.

## LLKs as Part of Tenstorrent's Software Stack

> LLKs are a foundational part of Tenstorrent’s software ecosystem. Apart from being the foundational layer, which allows for standalone software development, they are designed to work seamlessly with other tools such as:

Reviewer: ```suggestion
LLKs are a core part of Tenstorrent’s software ecosystem. In addition to enabling standalone software development, they are designed to work seamlessly with other tools such as:
```
Trying to avoid repetition (foundational x2)

> - [**TT-Metalium™**](https://github.com/tenstorrent/tt-metal): An open-source SDK that facilitates the development of custom kernels, offering low-level access to Tenstorrent’s hardware and leveraging LLKs for optimized performance.

Reviewer: ```suggestion
- [**TT-Metalium™**](https://github.com/tenstorrent/tt-metal): An open-source SDK that facilitates the development of custom kernels, offering low-level access to Tenstorrent hardware and leveraging LLKs for optimized performance.
```

## LLKs and the Tensix Core

LLKs can be categorized into compute and data movement kernels. Data movement kernels move data in and out of a Tensix core's internal memory, while compute kernels operate on data within the core. In this way, LLKs provide developers with a direct, optimized pathway to program the hardware, ensuring that every operation is executed with minimal overhead and maximum efficiency.

![Kernels to RISC Cores](images/LLK-L1-Tensix-2-25_official.png)

## LLKs as Part of Tenstorrent's Software Stack

LLKs are a foundational part of Tenstorrent’s software ecosystem. Apart from being the lowest layer, which allows for standalone software development, they are designed to work seamlessly with other tools such as:

- [**TT-Metalium™**](https://github.com/tenstorrent/tt-metal): An open-source SDK that facilitates the development of custom kernels, offering low-level access to Tenstorrent’s hardware and leveraging LLKs for optimized performance.
- [**TT-Forge™**](https://github.com/tenstorrent/tt-forge): A compiler that bridges machine learning frameworks with Tenstorrent hardware, using LLKs to optimize the execution of high-level models on Tensix cores.

> Together, these tools enable developers to build and optimize machine learning models and other high-performance applications tailored to Tenstorrent's hardware architecture.

Reviewer 2: Perhaps `these tools aim to enable...`. Since it doesn't work yet.

Author: Accepted.

![LLK in Tenstorrent Stack](images/LLK-L1-Tensix-3-25_official.png)

## Development and Testing

> The LLK repository provides the necessary resources for developing and testing kernels. It also includes a testing environment which validates that the LLKs function correctly on platforms like Wormhole and Blackhole. By targeting a single Tensix core, LLKs are optimized for high-performance execution under real-world workloads, ensuring that developers can fully leverage the hardware’s capabilities. The LLK repository has its own independent CI pipeline, enabling easier developer contributions.

Reviewer 2: Maybe ...its own independent CI **and test infrastructure**...
